File: algorithm/JavaScript/SerializeAndDeserializeBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

 function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
 }

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
 const serialize = function(root) {
    return serializeTree(root, '')
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
const deserialize = function(data) {
    const arr = data.split(',');
    return deserializeTree(arr)
};

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */

const serializeTree = function (root, res) {
    if (root === null) {
        res += 'None';
    } else {
        res += root.val + ',';
        res = serializeTree(root.left, res);
        res = serializeTree(root.right, res);
    }
    return res
}

const deserializeTree = function (arr) {
    if (arr[0] === 'None') {
        arr.shift();
        return null;
    }

    const root = new TreeNode(parseInt(arr[0]));
    arr.shift();
    root.left = deserializeTree(arr);
    root.right = deserializeTree(arr);

    return root

}

File: algorithm/python/SerializeAndDeserializeBinaryTree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        return self.serializeTree(root, '')
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        arr = data.split(',')
        return self.deserializeTree(arr)
    
    def serializeTree(self, root, res):
        if not root:
            res += 'None,'
        else:
            res += str(root.val) + ','
            res = self.serializeTree(root.left, res)
            res = self.serializeTree(root.right, res)
        
        return res
    
    def deserializeTree(self, arr):
        if arr[0] == 'None':
            arr.pop(0)
            return

        root = TreeNode(int(arr[0]))
        arr.pop(0)
        root.left = self.deserializeTree(arr)
        root.right = self.deserializeTree(arr)
        
        return root